panther_analysis_tool/lookup_table.py

```diff
@@ -0,0 +1,283 @@
+"""
+Panther Analysis Tool is a command line interface for writing,
+testing, and packaging policies/rules.
+Copyright (C) 2020 Panther Labs Inc
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as
+published by the Free Software Foundation, either version 3 of the
+License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import os
+import hashlib
+import json
+import logging
+from urllib import parse
+
+import boto3
+
+from panther_analysis_tool.util import get_client
+
+LOOKUP_LAMBDA_NAME = "panther-lookup-tables-api"
+LOG_PROCESSOR_LAMBDA_NAME = "panther-log-processor"
+UPLOAD_PREFIX = "luts"
+
+class LookupTable():
+    """A Panther Lookup Table"""
+
+    def __init__(self, spec: dict, spec_dir: str, aws_profile: str):
+        """Instantiate a Lookup Table object
+
+        Args:
+            spec: a dict representing the configuration of a Lookup Table
+        """
+        self.name = spec["name"]
+        self.spec = spec
+        self.spec_dir = spec_dir
+        self.aws_profile = aws_profile
+        self.client = get_client(self.aws_profile, "lambda")
+        self.lookup_id = None
+        self.input_bucket = ""
+        logging.info("Creating/updating the lookup table %s", self.name)
+
+
+    def create_update(self) -> int:
+        """Create a new Lookup Table or update a pre-existing one in Panther
+
+        Returns 1 if the operation fails
+
+        Args: none
+
+        Returns:
+        An integer return code
+        """
+        logging.info("Creating or updating the lookup table specified in %s", self.spec["dataFile"])
+        try:
+            self.input_bucket = self.get_input_bucket()
+        except Exception as err:  # pylint: disable=broad-except
+            logging.warning("\t[ERROR] Error fetching the input bucket name:\n%s", err)
+            return 1
+        if self.exists():
+            logging.info("Found the Lookup Table %s, updating it now", self.name)
+            return self.update()
+        logging.info("The Lookup Table %s does not exist, creating it now", self.name)
+        return self.create()
+
+
+    # check to see if a LUT with this name already exists (names are unique)
```

Comment:
You could reduce this code by adding an endpoint to the LUTs API called `createOrUpdate` that implements this logic. The Go code would be shorter also.

Reply:
OK, will move ahead on that front.

Reply:
done